### PR TITLE
[Go] Initial impelmentation of Go in new frontend

### PR DESCRIPTION
**Pull request by arthurscchan**
This is a working PR on adding the logic for go language in the new FI frontend.

**This PR is not ready for review**

### CODE REVIEW

**Comment by arthurscchan** [comment]
Here is a sample callgraph for fuzz_test.go in project clock.
```
FuzzClock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/fuzz_test.go -1
  time.ParseDuration ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/fuzz_test.go 25
  time.Parse ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/fuzz_test.go 29
  time.ParseDuration ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/fuzz_test.go 33
  NewMock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/fuzz_test.go 37
    time.Unix ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 88
  Add ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/fuzz_test.go 38
    Lock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 95
    Add ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 96
    Unlock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 97
    runNextTimer ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 101
      Lock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 155
      sort.Sort ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 158
      len ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 161
      Unlock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 162
      Next ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 168
      After ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 168
        Timer ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 185
          Lock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 252
          make ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 253
          Add ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 258
          append ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 261
          Unlock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 263
          runNextTimer ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 264
      Unlock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 169
      Next ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 174
      Unlock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 176
      Tick ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 179
        time.Tick ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 54
    Lock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 107
    Unlock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 109
    gosched ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 112
      time.Sleep ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 417
  Set ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/fuzz_test.go 39
    runNextTimer ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 120
    Lock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 126
    Unlock ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 128
    gosched ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/clock.go 131
  Timer ../../../oss_fuzz_integration/oss-fuzz/build/out/clock/src/clock/fuzz_test.go 40
```

File: src/fuzz_introspector/frontends/src/frontend_go.py [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+        if self.callsites:
+            return self.callsites
+
+        callsites = []
+        call_query = self.tree_sitter_lang.query('( call_expression ) @ce')
+        call_res = call_query.captures(self.root)
+        for _, call_exprs in call_res.items():
+            for call_expr in call_exprs:
+                for call_child in call_expr.children:
+                    # Ignore testing.F.Fuzz
```
Comment (DavidKorczynski) [inline src/fuzz_introspector/frontends/src/frontend_go.py:548]: Why are we ignoring this one?

File: src/fuzz_introspector/frontends/src/frontend_go.py [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+
+class FunctionMethod():
+    """Wrapper for a General Declaration for function/method"""
+
+    def __init__(self, root: Node, tree_sitter_lang: Language,
+                 source_code: SourceCodeFile, is_function: bool):
+        self.root = root
+        self.tree_sitter_lang = tree_sitter_lang
+        self.parent_source = source_code
+        self.is_function = is_function
```
Comment (DavidKorczynski) [inline src/fuzz_introspector/frontends/src/frontend_go.py:315]: what is `is_function` used for?

File: src/fuzz_introspector/frontends/src/frontend_go.py [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+
+        function_res = func_query.captures(self.root)
+        for _, funcs in function_res.items():
+            for func in funcs:
+                self.functions.append(
+                    FunctionMethod(func, self.tree_sitter_lang, self, True))
+
+    def _set_method_declaration(self):
+        """Internal helper for retrieving all methods."""
+        func_query_str = '( method_declaration ) @fd '
```
Comment (DavidKorczynski) [inline src/fuzz_introspector/frontends/src/frontend_go.py:86]: what is the difference between `method_declaration` and `function_declaration`?

File: src/fuzz_introspector/frontends/src/frontend_go.py [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+                imports.append(match.group('single'))
+            elif match.group('multi'):
+                multi = [
+                    line.strip().strip('"')
+                    for line in match.group('multi').splitlines()
+                ]
+                imports.extend(multi)
+
+        self.imports = [
+            s.rsplit('/', 1)[-1] if '/' in s else s for s in imports
```
Comment (DavidKorczynski) [inline src/fuzz_introspector/frontends/src/frontend_go.py:113]: could you leave a comment on this? What does it do

File: src/fuzz_introspector/frontends/src/frontend_go.py [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+
+        function_res = func_query.captures(self.root)
+        for _, funcs in function_res.items():
+            for func in funcs:
+                self.methods.append(
+                    FunctionMethod(func, self.tree_sitter_lang, self, False))
+
+    def _set_imports(self):
+        """Internal helper for retrieving all imports."""
+        import_pattern = r'^\s*import\s+(?:(?P<single>"[^"]+")|\((?P<multi>[^)]+)\))'
```
Comment (DavidKorczynski) [inline src/fuzz_introspector/frontends/src/frontend_go.py:97]: Do we really need to look at the source code for this? Can't we traverse the syntax tree?

Just looked at the tree for simple program:

```go
package main

import "fmt"

import "rsc.io/quote"

func main() {
    fmt.Println(quote.Go())
}
```
gives the tree:
```
source_file [0, 0] - [9, 0]
  package_clause [0, 0] - [0, 12]
    package_identifier [0, 8] - [0, 12]
  import_declaration [2, 0] - [2, 12]
    import_spec [2, 7] - [2, 12]
      path: interpreted_string_literal [2, 7] - [2, 12]
  import_declaration [4, 0] - [4, 21]
    import_spec [4, 7] - [4, 21]
      path: interpreted_string_literal [4, 7] - [4, 21]
      ...
```

We should be able to traverse `import_declaration`?

Sample logic in C: https://github.com/ossf/fuzz-introspector/blob/60cb9940908f1ab6a4744e3e218baafe0393a618/src/fuzz_introspector/frontends/src/frontend_c.py#L490-L503

**Review by DavidKorczynski — CHANGES_REQUESTED** [review]
Looks good -- some minor comments

File: src/fuzz_introspector/frontends/src/frontend_go.py (hunk 1 above)
Comment (arthurscchan) [reply src/fuzz_introspector/frontends/src/frontend_go.py]: Forget to remove. Now removed.

File: src/fuzz_introspector/frontends/src/frontend_go.py (hunk 4 above)
Comment (arthurscchan) [reply src/fuzz_introspector/frontends/src/frontend_go.py]: Done so.

File: src/fuzz_introspector/frontends/src/frontend_go.py (hunk 2 above)
Comment (arthurscchan) [reply src/fuzz_introspector/frontends/src/frontend_go.py:314]: it is used to distinguish functions from methods. But currently it is just a unused flag. I am storing this in order for future reference if some process needed to identify them.

File: src/fuzz_introspector/frontends/src/frontend_go.py (hunk 3 above)
Comment (arthurscchan) [reply src/fuzz_introspector/frontends/src/frontend_go.py:86]: From go tree definition, it contains both method and function declaration. Which both of them needed to be handled in order to covers all of the functions and methods.
```
(
  (comment)* @doc
  .
  (function_declaration
    name: (identifier) @name) @definition.function
  (#strip! @doc "^//\\s*")
  (#set-adjacent! @doc @definition.function)
)

(
  (comment)* @doc
  .
  (method_declaration
    name: (field_identifier) @name) @definition.method
  (#strip! @doc "^//\\s*")
  (#set-adjacent! @doc @definition.method)
)
```

File: src/fuzz_introspector/frontends/src/frontend_go.py (hunk 5 above)
Comment (arthurscchan) [reply src/fuzz_introspector/frontends/src/frontend_go.py]: Have changed.

**Review by DavidKorczynski — APPROVED** [review]

**Pull request merged**